examples/extra_element_integer/clustering_algorithm.cpp

```diff
@@ -50,54 +50,78 @@ void AddVariablesToSystem(libMesh::EquationSystems &equation_systems,
+  for (const auto &elem : mesh.element_ptr_range()) {
+    elem->set_extra_integer(index, elem->id());
+  }
+}
 
-  const unsigned int variable_num = system.variable_number(variable_name);
+bool BelongToCluster(libMesh::Elem *elem, libMesh::Elem *neighbor_elem,
+                     int variable_num, libMesh::LinearImplicitSystem &system) {
 
-  const unsigned int index = mesh.add_elem_integer(variable_name);
+  double element_solution = GetElementDataFromMesh(system, elem, variable_num);
+  double neighbor_solution =
+      GetElementDataFromMesh(system, neighbor_elem, variable_num);
+  if (static_cast<int>(element_solution) ==
+      static_cast<int>(neighbor_solution)) {
 
-  for (const auto &elem : mesh.element_ptr_range()) {
-
-    int element_solution =
-        static_cast<int>(GetElementDataFromMesh(system, *elem, variable_num));
+    return true;
+  } else {
+    return false;
+  }
+}
 
-    bool belong_to_a_cluster = false;
-    for (unsigned int side = 0; side < elem->n_sides(); ++side) {
-      const Elem *neighbor = elem->neighbor_ptr(side);
-      if (neighbor) {
+const unsigned int FindCluster(libMesh::Mesh &mesh,
+                               libMesh::LinearImplicitSystem &system,
+                               const std::string &variable_name) {
+  const unsigned int variable_num = system.variable_number(variable_name);
+  const unsigned int index = mesh.add_elem_integer(variable_name);
 
-        int neighbor_element_solution = static_cast<int>(
-            GetElementDataFromMesh(system, *neighbor, variable_num));
+  SetElementIDasClusterID(mesh, index);
 
-        if (element_solution == neighbor_element_solution) {
-          belong_to_a_cluster = true;
-          // Danger Zone for future ref
-          // cluster_id should never be equal to element_solution
-          unsigned int cluster_id = element_solution;
-          elem->set_extra_integer(index, cluster_id);
-          break;
+  for (const auto &elem : mesh.element_ptr_range()) {
+    if (elem->get_extra_integer(index) == elem->id()) {
+
+      std::vector<libMesh::Elem *> neighbor_list;
+      neighbor_list.push_back(elem);
+
+      while (neighbor_list.size() > 0) {
+        libMesh::Elem *current_elem = neighbor_list.back();
+        neighbor_list.pop_back();
+
+        for (unsigned int s = 0; s < current_elem->n_sides(); s++) {
+          libMesh::Elem *neighbor_elem = current_elem->neighbor_ptr(s);
+
+          if (neighbor_elem &&
+              neighbor_elem->get_extra_integer(index) != elem->id() &&
+              (BelongToCluster(current_elem, neighbor_elem, variable_num,
+                               system))) {
+            neighbor_elem->set_extra_integer(index, elem->id());
+            std::cout << "new element to the cluster = " << neighbor_elem->id()
+                      << " derived from = " << current_elem->id() << std::endl;
+            neighbor_list.push_back(neighbor_elem);
+          }
         }
       }
```

Comment:
If you initialize the `extra_integer` with some value `not_visited` (e.g. -1)

```suggestion
  std::stack<libMesh::Elem *> neighbor_stack;
  libMesh::Elem* test_elem, neighbor_elem;
  
  for (const auto &elem : mesh.element_ptr_range()) {
    neighbor_stack.push(elem);
    int cluster_id = elem->id();
    
    while (!neighbor_stack.empty()) {
      test_elem = neighbor_stack.top();
      neighbor_stack.pop();

      if (test_elem->get_extra_integer(index) == not_visited &&
          BelongToCluster(elem, test_elem, variable_num, system) ) {
          
          test_elem->set_extra_integer(index, cluster_id);
          
          for (unsigned int n = 0; n < test_elem->n_sides(); n++) {
            neighbor_elem = current_elem->neighbor_ptr(n);
            if (neighbor_elem) neighbor_stack.push(neighbor_elem);
              
          }
       }
     }
```

Reply:
Note that for each element that you loop through, you push it onto the stack to start, and then immediately pop it off the stack again.  But by doing this, that element is treated the same way as all others.

Reply:
If I am guessing right it should also set the extra_integer as element_id if that element doesn't belong to any cluster?